lib/shelley/src/Cardano/Wallet/Shelley/Network.hs

```diff
@@ -0,0 +1,508 @@
+                            clientStIdle
+                    }
+
+            ReqNetworkTip respond -> do
+                -- Alternatively, we could simply 'SendMsgRequestNext' which
+                -- also returns a tip and is probably less expensive on the
+                -- node's end.
+                pure $ SendMsgFindIntersect [] $ ClientStIntersect
+                    { recvMsgIntersectImproved = \_intersection tip ->
+                        ChainSyncClient $ respond tip *> clientStIdle
+                    , recvMsgIntersectUnchanged = \tip ->
+                        ChainSyncClient $ respond tip *> clientStIdle
+                    }
+
+        -- 'Next' state coming from a 'ReqNextBlocks' request. We stay in that
+        -- state for a couple of messages, accumulating blocks as we go and
+        -- returning after a certain number has been fetched.
+        --
+        -- Note that the process is interrupted prematurely if there's no more
+        -- block to fetched (caught up with the tip). Also, we consider
+        -- rollbacks as an error since the wallet engine can't handle them at
+        -- the moment.
+        clientStFetchingBlocks
+            :: Int
+                -- ^ Number of messages to fetch
+            -> Point
+                -- ^ Starting point
+            -> [Block]
+                -- ^ Block accumulator
+            -> (Either ErrNextBlocks [Block] -> m ())
+                -- ^ Success or Error callback
+            -> m (ClientStIdle Block Point m Void)
+        clientStFetchingBlocks 0 _ blocks respond = do
+            respond (Right blocks)
+            clientStIdle
+        clientStFetchingBlocks n start blocks respond = pure $ SendMsgRequestNext
+            (ClientStNext
+                { recvMsgRollForward = \block _tip -> ChainSyncClient $
+                    clientStFetchingBlocks (n-1) start (block:blocks) respond
+
+                , recvMsgRollBackward = \point _tip -> ChainSyncClient $
+                    -- NOTE
+                    -- The server always start by asking the client to rollback
+                    -- to the intersection point. Other form of rollbacks aren't
+                    -- _yet_ expected but will eventually have to be handled.
+                    -- For now, we consider this to be an error.
+                    if point == start
+                        then clientStFetchingBlocks n start [] respond
+                        else do
+                            respond (Left ErrNextBlocksRollBack)
+                            clientStIdle
+                }
+            )
+            (do
+                -- NOTE
+                -- In the 'Next' state, a client can receive either an immediate
+                -- response (previous handler), or a response indicating that
+                -- the blocks can't be retrieve _on the moment_ but will be
+                -- after a while. In this case, we simply resume the response
```

Comment:
> resume the response

I'm not sure I understand this part.

Reply:
I'll try to clarify the comment here: When sending a `MsgRequestNext`, we have to provide two callback functions:

- One for immediate response, in case where the node has a corresponding block to send
- One for delayed response, in case where we are already at the tip and the next block hasn't been yet 

In the latte case, we actually get an immediate response from the node telling us that "data will arrive soon" and give us the ability to execute some actions in the meantime. In this implementation, I chose to "abort" the handler by immediately sending a response to our network client. The `respond` identifier here refers to a callback function for the client of the network layer. It prevents hanging for a whole slot-length while querying the next blocks.